File: app/src/main/java/com/vmenon/mpo/core/MPOMediaPlayer.java
package com.vmenon.mpo.core;

import android.media.MediaPlayer;
import android.util.Log;
import android.view.SurfaceHolder;

import java.io.File;
import java.io.FileDescriptor;
import java.io.FileInputStream;
import java.io.IOException;

/**
 * Wrapper around actual media player mechanism (i.e. {@link android.media.MediaPlayer})
 */
public class MPOMediaPlayer extends MPOPlayer implements MediaPlayer.OnPreparedListener,
        MediaPlayer.OnErrorListener,
        MediaPlayer.OnCompletionListener, MediaPlayer.OnSeekCompleteListener {

    private MediaPlayer mediaPlayer;
    private long currentPosition;
    private SurfaceHolder surfaceHolder;

    public MPOMediaPlayer() {

    }

    @Override
    protected void doPrepareForPlayback(File file) {
        createMediaPlayerIfNeeded();
        FileInputStream fis = null;
        try {
            fis = new FileInputStream(file);
            FileDescriptor fd = fis.getFD();
            mediaPlayer.setDataSource(fd);
            mediaPlayer.prepareAsync();
        } catch (IOException e) {
            Log.w("Can't play music", e);
        } finally {
            if (fis != null) {
                try {
                    fis.close();
                } catch (IOException e) {
                }
            }
        }
    }

    @Override
    public void play() {
        mediaPlayer.start();
    }

    @Override
    public void pause() {
        if (mediaPlayer != null && mediaPlayer.isPlaying()) {
            mediaPlayer.pause();
            currentPosition = mediaPlayer.getCurrentPosition();
        }
    }

    @Override
    public void stop() {
        if (mediaPlayer != null) {
            currentPosition = getCurrentPosition();
            mediaPlayer.stop();
        }
    }

    @Override
    public long getCurrentPosition() {
        return mediaPlayer != null ? mediaPlayer.getCurrentPosition() : currentPosition;
    }

    @Override
    public boolean isPlaying() {
        return mediaPlayer != null && mediaPlayer.isPlaying();
    }

    @Override
    public void seekTo(long position) {
        Log.d("MPO", "seekTo called with " + position);

        if (mediaPlayer == null) {
            // If we do not have a current media player, simply update the current position
            currentPosition = position;
        } else {
            mediaPlayer.seekTo((int) position);
        }
    }

    @Override
    public void setVolume(float volume) {
        mediaPlayer.setVolume(volume, volume);
    }

    @Override
    public void setDisplay(SurfaceHolder surfaceHolder) {
        this.surfaceHolder = surfaceHolder;
        if (mediaPlayer != null) {
            mediaPlayer.setDisplay(surfaceHolder);
        }
    }

    @Override
    protected void doCleanUp() {
        if (mediaPlayer != null) {
            mediaPlayer.reset();
            mediaPlayer.release();
            mediaPlayer = null;
        }
    }

    @Override
    public void onPrepared(MediaPlayer mediaPlayer) {
        if (listener != null) {
            listener.onMediaPrepared();
        }
    }

    @Override
    public boolean onError(MediaPlayer mediaPlayer, int i, int i1) {
        return false;
    }

    @Override
    public void onCompletion(MediaPlayer mediaPlayer) {
        if (listener != null) {
            listener.onMediaFinished();
        }
    }

    @Override
    public void onSeekComplete(MediaPlayer mediaPlayer) {
        mediaPlayer.start();
        if (listener != null) {
            listener.onMediaSeekFinished();
        }
    }

    private void createMediaPlayerIfNeeded() {
        Log.d("MPO", "createMediaPlayerIfNeeded. needed? " + (mediaPlayer == null));
        if (mediaPlayer == null) {
            mediaPlayer = new MediaPlayer();

            /** TODO
             // Make sure the media player will acquire a wake-lock while
             // playing. If we don't do that, the CPU might go to sleep while the
             // song is playing, causing playback to stop.
             mMediaPlayer.setWakeMode(mService.getApplicationContext(),
             PowerManager.PARTIAL_WAKE_LOCK);*/

            // we want the media player to notify us when it's ready preparing,
            // and when it's done playing:
            mediaPlayer.setOnPreparedListener(this);
            mediaPlayer.setOnCompletionListener(this);
            mediaPlayer.setOnErrorListener(this);
            mediaPlayer.setOnSeekCompleteListener(this);
            if (surfaceHolder != null) {
                mediaPlayer.setDisplay(surfaceHolder);
            }
        } else {
            mediaPlayer.reset();
        }
    }
}

File: player-framework/src/main/java/com/vmenon/mpo/player/framework/MPOMediaPlayer.kt
package com.vmenon.mpo.player.framework

import android.media.MediaPlayer
import android.util.Log
import android.view.SurfaceHolder

import java.io.File
import java.io.FileInputStream
import java.io.IOException

/**
 * Wrapper around actual media player mechanism (i.e. [android.media.MediaPlayer])
 */
class MPOMediaPlayer : BaseMPOPlayer(), MediaPlayer.OnPreparedListener, MediaPlayer.OnErrorListener,
    MediaPlayer.OnCompletionListener, MediaPlayer.OnSeekCompleteListener {

    private var mediaPlayer: MediaPlayer? = null
    private var currentPosition: Long = 0
    private var surfaceHolder: SurfaceHolder? = null

    override val isPlaying: Boolean
        get() = mediaPlayer?.isPlaying == true

    override fun doPrepareForPlayback(file: File) {
        createMediaPlayerIfNeeded()
        mediaPlayer?.let { player ->
            var fis: FileInputStream? = null
            try {
                fis = FileInputStream(file)
                val fd = fis.fd
                player.setDataSource(fd)
                player.prepareAsync()
            } catch (e: IOException) {
                Log.w("Can't play music", e)
            } finally {
                if (fis != null) {
                    try {
                        fis.close()
                    } catch (e: IOException) {
                    }

                }
            }
        }

    }

    override fun play() {
        mediaPlayer?.start()
    }

    override fun pause() {
        mediaPlayer?.let { player ->
            if (player.isPlaying) {
                player.pause()
                currentPosition = player.currentPosition.toLong()
            }
        }
    }

    override fun stop() {
        currentPosition = getCurrentPosition()
        mediaPlayer?.stop()
    }

    override fun getCurrentPosition(): Long {
        return mediaPlayer?.currentPosition?.toLong() ?: currentPosition
    }

    override fun seekTo(position: Long) {
        Log.d("MPO", "seekTo called with $position")

        if (mediaPlayer == null) {
            // If we do not have a current media player, simply update the current position
            currentPosition = position
        } else {
            mediaPlayer?.seekTo(position.toInt())
        }
    }

    override fun setVolume(volume: Float) {
        mediaPlayer?.setVolume(volume, volume)
    }

    override fun setDisplay(surfaceHolder: SurfaceHolder?) {
        this.surfaceHolder = surfaceHolder
        mediaPlayer?.setDisplay(surfaceHolder)
    }

    override fun doCleanUp() {
        mediaPlayer?.let { player ->
            player.reset()
            player.release()
        }
        mediaPlayer = null
    }

    override fun onPrepared(mediaPlayer: MediaPlayer) {
        mListener?.onMediaPrepared()
    }

    override fun onError(mediaPlayer: MediaPlayer, i: Int, i1: Int): Boolean {
        return false
    }

    override fun onCompletion(mediaPlayer: MediaPlayer) {
        mListener?.onMediaFinished()
    }

    override fun onSeekComplete(mediaPlayer: MediaPlayer) {
        mediaPlayer.start()
        mListener?.onMediaSeekFinished()
    }

    private fun createMediaPlayerIfNeeded() {
        Log.d("MPO", "createMediaPlayerIfNeeded. needed? " + (mediaPlayer == null))
        if (mediaPlayer == null) {
            mediaPlayer = MediaPlayer()

            /** TODO
             * // Make sure the media player will acquire a wake-lock while
             * // playing. If we don't do that, the CPU might go to sleep while the
             * // song is playing, causing playback to stop.
             * mMediaPlayer.setWakeMode(mService.getApplicationContext(),
             * PowerManager.PARTIAL_WAKE_LOCK); */

            // we want the media player to notify us when it's ready preparing,
            // and when it's done playing:
            mediaPlayer?.setOnPreparedListener(this)
            mediaPlayer?.setOnCompletionListener(this)
            mediaPlayer?.setOnErrorListener(this)
            mediaPlayer?.setOnSeekCompleteListener(this)
            surfaceHolder?.let {
                mediaPlayer?.setDisplay(it)
            }
        } else {
            mediaPlayer?.reset()
        }
    }
}
